[A11Y] Add focus traps to modals and nav drawer

davwheat opened the pull request:
<!--
IMPORTANT: We applaud pull requests, they excite us every single time. As we have an obligation to maintain a healthy code standard and quality, we take sufficient time for reviews. Please do create a separate pull request per change/issue/feature; we will ask you to split bundled pull requests.
-->

**Fixes #2663**
**Fixes #2665**

> ### **Please merge #3007 before this PR!**

**Changes proposed in this pull request:**
<!-- fill this out, mention the pages and/or components which have been impacted -->
- Add focus trap to modals
- Add focus trap to drawer on phone viewports
- Auto close drawer when leaving phone viewport
- Replace deprecated JQuer (`$.click()` -> `$.on('click')`)
- Use `requestAnimationFrame` instead of `setTimeout`

**Reviewers should focus on:**
<!-- fill this out, ask for feedback on specific changes you are unsure about -->

**Screenshot**
<!-- include an image of the most relevant user-facing change, if any -->

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.

askvortsov1 commented on js/src/common/components/ModalManager.js:
```diff
@@ -1,40 +1,72 @@
+        if (this.modalShown) this.focusTrap.activate?.();
+        else this.focusTrap.deactivate?.();
+      } catch {
+        // We can expect errors to occur here due to the nature of mithril rendering
```
Could we catch a more specific kind of exception?

davwheat replied:
I thought about that, but thought the added code needed to check if the exception was the one we'd expect wasn't worth it.

JS doesn't have error classes like PHP and other languages, meaning the only way to catch this specific error would be comparing the error message against a constant.

![image](https://user-images.githubusercontent.com/7406822/129459727-a9a2da0f-6b0a-476a-a2da-2e944a66306f.png)

askvortsov1 replied:
> JS doesn't have error classes like PHP and other languages

Never thought I'd see you admit an upside of PHP :wink:

davwheat replied:
Never ever mention it again. 😄

datitisev replied:
Hm I don't like the idea of catching all errors and not just the one we're trying to avoid spamming...

askvortsov1 commented on js/src/common/components/ModalManager.js:
```diff
@@ -1,40 +1,72 @@
   animateShow(readyCallback) {
     const dismissible = !!this.attrs.state.modal.componentClass.isDismissible;
 
+    this.modalShown = true;
```
Maybe these should happen before the jquery calls? That way, if there's rerendering triggered by jquery (is that a thing?), it'll be applied correctly.

davwheat replied:
I'm not sure I understand what you're saying. Are you referring to `this.modalShown = false;` being called after removing the `in` class?

jQuery shouldn't ever cause a redraw, unless you're using it to call a redraw of course, or after any interaction event is fired , and subsequently handled, on a mithril-mounted DOM node.

askvortsov1 commented on js/src/common/utils/Drawer.js:
```diff
@@ -1,20 +1,41 @@
+
+  resizeHandler(e) {
+    if (this.isOpen()) {
+      if (document.documentElement.style.getPropertyValue('--flarum-screen') !== 'phone') {
```
Can we use the `app.screen` encapsulation?

davwheat replied:
Ooh, I didn't realise that existed! Thanks for pointing it out!

askvortsov1 commented on js/src/common/utils/Drawer.js:
```diff
@@ -1,20 +1,41 @@
     // Set up an event handler so that whenever the content area is tapped,
     // the drawer will close.
-    $('#content').click((e) => {
+    $('#content').on('click', (e) => {
```
Maybe we should just use the native JS method?

askvortsov1 commented on js/src/common/utils/Drawer.js:
```diff
@@ -1,20 +1,41 @@
+  }
+
+  resizeHandler(e) {
+    if (this.isOpen()) {
```
Can we use `&&` instead of nesting?

askvortsov1 commented on js/src/common/utils/Drawer.js:
```diff
@@ -1,20 +1,41 @@
+    this.focusTrap = createFocusTrap('#drawer', { allowOutsideClick: true });
+  }
+
+  resizeHandler(e) {
```
When are these resizes actually an issue? I can imagine changes in height easily (keyboard open), but why would we have changes in width?

davwheat replied:
See: https://github.com/flarum/core/pull/3018/commits/438577dae7a79f78cd2e6481d112340af5870917

The focus would end up staying trapped in the header elements if the drawer was opened, then the window was resized to be larger. There'd then be no way to close the drawer (and thus escape the focus trap) without making the window smaller again.

davwheat replied:
I discovered, thanks to a kind person in `#extend` in Discord, this: https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/onchange

No need for resize handlers! We can detect when a media query changes state in JS!

```ts
const phoneMaxWidth = document.documentElement.getPropertyValue('--screen-phone-max')
const drawerAailableMediaQuery = window.matchMedia(`(max-width: ${phoneMaxWidth})`);

drawerAailableMediaQuery.addEventListener('change', (e) => {
  if (e.matches) {
    // Is phone
  } else {
    // Is not phone
  }
})
```

askvortsov1 replied:
Isnt that also effectively a resize handler, but one that only responds to the exact change we care about?

askvortsov1 commented on js/src/common/utils/Drawer.js:
```diff
@@ -41,7 +62,10 @@ export default class Drawer {
 
-    if (!$app.hasClass('drawerOpen')) return;
+    this.focusTrap.deactivate();
+    window.removeEventListener('resize', this.throttledResizeHandler);
```
Are we calling this onremove too? Or just on hide?

davwheat replied:
I think only on hide should be fine. The Drawer class isn't a component, so we don't have access to the `onremove` lifecycle hook here, and, as such, this class never gets "recreated" in the same way mithril components can.

askvortsov1 commented on js/src/common/utils/Drawer.js:
```diff
@@ -50,7 +74,7 @@ export default class Drawer {
     $app.removeClass('drawerOpen');
 
-    if (this.$backdrop) this.$backdrop.remove();
+    this.$backdrop.remove?.();
```
Nice

datitisev replied:
This changes the logic. You want to check for $backdrop, the current code checks that the `remove` button is truthy, which will error if `$backdrop` is undefined

![image](https://user-images.githubusercontent.com/6401250/129460003-8a86c7e7-ef30-43ab-9129-2d87e7020974.png)

davwheat replied:
Now I look at it, why do we need this at all?

From what I can see, after you remove the element with jQuery, the variable doesn't get cleared.

The only scenario would be if `hide` was somehow called before `show`. Is that even possible?

```js
let x = $('<div />').appendTo('body')
!!x // true
x.remove()
!!x // true
x.remove()
// ...
```

datitisev replied:
I'd still keep the check just in case. 🤷 No reason not to have it, right?

davwheat replied:
True. I've combined the two as if the value is anything other than `undefined`, we get an error if we only have a `?` after `$backdrop.`

davwheat commented on js/src/common/components/ModalManager.js:
```diff
@@ -1,40 +1,72 @@
+import { createFocusTrap } from '../utils/focusTrap';
+import { tabbable } from 'tabbable';
+
+window.tabbable = tabbable;
```
No, this was from debugging -- I just totally forgot to remove it.

davwheat commented on js/src/common/components/ModalManager.js:
```diff
@@ -1,40 +1,72 @@
-              state: this.attrs.state,
-            })
-          : ''}
+        {!!modal &&
```
No, `!undefined === true`, meaning that `!!undefined` would be `false`.

Mithril will attempt to render the result of `&&` as a text node if the left hand side resolves to false but is not actually a boolean.

For example:
```js
return <>{0 && <MyElement />}</>
```

Would render a '0' into the DOM.

I've used `!!` here to maintain consistency with other areas of core where we use `&&` to conditionally render components. I've been swapping out ternaries with this method as it's slightly more efficient.

davwheat commented on js/package.json:
```diff
@@ -10,6 +10,7 @@
     "color-thief-browser": "^2.0.2",
     "dayjs": "^1.10.4",
     "expose-loader": "^1.0.3",
+    "focus-trap": "^6.6.0",
```
Seems to be fine for me:

![image](https://user-images.githubusercontent.com/7406822/129458868-5f11915c-be96-425c-a89d-9353c36ccf9a.png)

davwheat commented on js/src/common/components/ModalManager.js:
```diff
@@ -1,40 +1,72 @@
   }
 
+  /**
+   * @param {import('mithril').VnodeDOM<{}, this>} vnode
```
I'm a little ahead of you on this one! I made a TS conversion PR a couple of days ago :P

https://github.com/flarum/core/pull/3007

datitisev commented on js/src/common/utils/Drawer.js:
```diff
@@ -1,28 +1,72 @@
+
+    this.appElement = document.getElementById('app');
+    this.focusTrap = createFocusTrap('#drawer', { allowOutsideClick: true });
+    this.drawerAailableMediaQuery = window.matchMedia(
```
```suggestion
    this.drawerAvailableMediaQuery = window.matchMedia(
```

datitisev commented on js/src/forum/utils/KeyboardNavigatable.ts:
```diff
@@ -51,20 +51,35 @@ export default class KeyboardNavigatable {
       callback(e);
     };
 
-    this.callbacks.set(9, handler);
```
This is technically a breaking change as it changes behavior, is it not? Not sure how we want to handle it, if at all.

davwheat replied:
It is, yes. We *could* add a second param to override the original behaviour, which might be a better idea.

datitisev replied:
I'm in favor of replacing the second param with a diff method for 2.0. Just want to void breaking outside of major releases.

datitisev reviewed (approved):
Not tested locally, but code looks good aside from a few comments.

datitisev reviewed (changes requested):
I didn't mean to approve.

davwheat commented:
YOU WILL GET CO-AUTHORSHIP OF THAT COMMIT IF IT IS THE LAST THING I DO

davwheat commented:
Rebase in progress...

@datitisev you still have "requested changes" marked on this PR -- I think I addressed your comment about a breaking change, but I'll re-request your review and you can either approve, dismiss the request, or request other changes.

davwheat commented on js/src/common/components/ModalManager.tsx:
```diff
@@ -13,7 +15,14 @@ interface IModalManagerAttrs {
  */
 export default class ModalManager extends Component<IModalManagerAttrs> {
-  view() {
+  protected focusTrap: FocusTrap | undefined;
+
+  /**
+   * Whether a modal is currently shown by this modal manager.
+   */
+  protected modalShown: boolean = false;
```
/cc reviewers @askvortsov1 @SychO9 

What are your opinions on the accessibility modifiers here? Should these be public, private or protected? I went with protected for now, as that gives the best of both worlds, but I do wonder if it should be public instead.

SychO9 replied:
I prefer it protected, I don't think we'd be breaking any extensions either.

davwheat replied:
The only thing I thought was if an extension closes a modal without using `animateHide()`, then `modalShown` won't be updated, and the focus trap will still exist and be active.

SychO9 replied:
how would that happen ?

davwheat replied:
If they directly called `$('.ModalManager').modal('hide')`, or whatever the right selector is. I can't actually think of any reason why they'd do it, so I'm fine with `protected`.

SychO9 replied:
yea, there are so many ways extension devs could do things the wrong way, I don't think we can ever do anything about that except have good docs explaining best practices.

askvortsov1 replied:
Let's keep it as protected. We offer a public API for closing the modal, anything else is explicitly going outside of that public API, and therefore not guarunteed not to break.
